charmhub_lp_tools/charm_project.py

```diff
@@ -154,47 +150,216 @@ def close(
+        found_branch = None
+        for branch, branch_data in self.project.branches.items():
+            for channel in branch_data['channels']:
+                if channel == find_channel:
+                    # found the branch, let's set it.
+                    if branch.startswith("refs/heads/"):
+                        found_branch = branch[len("refs/heads/"):]
+                    else:
+                        found_branch = branch
+                    # break the inner loop
+                    break
+            else:
+                # continue outer loop if the inner loop wasn't broken
+                continue
+            # Inner loop was broken, break outer loop
+            break
+        if found_branch is None:
+            self.log.debug("get_charm_metadata_for_channel: "
+                           "couldn't find branch, so returning None.")
+            return None
+        if repo.startswith("https://opendev.org/"):
+            # repo is https://opendev.org/openstack/charm-keystone.git
+            # url is https://opendev.org/{user}/{charm}/raw/branch/
+            #              {branch}/metadata.yaml
+            user_charm = repo[len("https://opendev.org/"):]
+            (user, charm) = user_charm.split('/')
+            if charm.endswith(".git"):
+                charm = charm[:-4]
+            url = (f"https://opendev.org/{user}/{charm}/raw/branch/"
+                   f"{found_branch}/metadata.yaml")
+            # use requests to get the metadata
+            try:
+                self.log.debug("Getting metadata from '%s'", url)
+                result = requests.get(url)
+                if result.status_code != 200:
+                    self.log.info("Error getting metadata, code=%s",
+                                  result.status_code)
+                    return None
+                result_text = result.text.strip()
+                if result_text == "src/metadata.yaml":
+                    url = (f"https://opendev.org/{user}/{charm}/raw/branch/"
+                           f"{found_branch}/src/metadata.yaml")
+                    self.log.debug("Following link: getting metadata from"
+                                   " '%s'", url)
+                    result = requests.get(url)
+                    result_text = result.text.strip()
+                try:
+                    return yaml.safe_load(result_text)
+                except Exception as e:
+                    self.log.error(
+                        "Couldn't decode response due to %s\nResponse:\n%s",
+                        str(e), result.text)
+            except Exception as e:
+                self.log.error('Failed to fetch metadata.yam: %s', str(e))
+                return None
+
+        else:
+            # don't know what to do with that one.
+            self.log.error("Don't know how to fetch metadata.yaml from "
+                           " %s", repo)
+            return None
```

Comment:
`None` is consistent with the rest of the function; i.e. it couldn't work out how to fetch the metadata, so from the function's perspective there isn't any; from the caller's perspective, there is no metadata and it has to deal with that.  It would also have to deal with the Exception, unless "not being able to determine metadata" is a bug and so should crash the program?